getting-started/installation/general-installation/installation-community-containerized.markdown

```diff
@@ -0,0 +1,286 @@
+    privileged: true
+    command:
+      - /bin/sh
+      - -c
+      - |
+        "/usr/local/sbin/cf-agent --bootstrap $(ip -4 -o addr show eth0 | awk '{print $4}' | cut -d'/' -f1)"
+    networks:
+      - control-plane
+
+  cfengine-host:
+    image: cfengine:lts
+    build:
+      context: .
+      dockerfile: Dockerfile
+    privileged: true
+    command:
+      - /bin/sh
+      - -c
+      - |
+        "/usr/local/sbin/cf-agent --bootstrap $(dig +short cfengine-hub|tr -d [:space:])"
+    networks:
+      - control-plane
+    depends_on:
+      cfengine-hub:
+        condition: service_healthy
+        required: true
+
+networks:
+  control-plane:
+```
+
+Validate the `compose.yaml` file
+
+```command
+docker compose -f compose.yaml config 1>/dev/null
+```
+**Note**: No output means valid yaml file.
+
+Start service cfengine-demo
+
+```command
+docker compose -f compose.yaml up -d
+```
+
+Bootstrap hub and hosts
+
+```command
+docker exec -it cfengine-hub bash -c "/usr/local/sbin/cf-agent --bootstrap \$(ip -4 -o addr show eth0 | awk '{print \$4}' | cut -d'/' -f1)"
+```
+```output
+R: Bootstrapping from host '192.168.16.2' via built-in policy '/var/cfengine/inputs/failsafe.cf'
+R: This host assumes the role of policy server
+R: Updated local policy from policy server
+R: Triggered an initial run of the policy
+R: Restarted systemd unit cfengine3
+notice: Bootstrap to '192.168.16.2' completed successfully!
+```
+
+```command
+docker exec -it cfengine-demo-cfengine-host-1 bash -c "/usr/local/sbin/cf-agent --bootstrap \$(dig +short cfengine-hub|tr -d [:space:])"
+```
+```output
+notice: Bootstrap mode: implicitly trusting server, use --trust-server=no if server trust is already established
+notice: Trusting new key: MD5=2f406e11cfd3e08d810d77a186e204e2
+R: Bootstrapping from host '192.168.16.2' via built-in policy '/var/cfengine/inputs/failsafe.cf'
+R: This autonomous node assumes the role of voluntary client
+R: Updated local policy from policy server
+R: Triggered an initial run of the policy
+R: Restarted systemd unit cfengine3
+notice: Bootstrap to '192.168.16.2' completed successfully!
+```
+
+Health-check for hub and host
+
+```command
+docker exec -it cfengine-hub bash -c "/usr/local/sbin/cf-agent --self-diagnostics"
+```
+```output
+...
+[ YES ] Check that agent is bootstrapped: 192.168.16.2
+[ YES ] Check if agent is acting as a policy server: Acting as a policy server
+[ YES ] Check private key: OK at '/var/cfengine/ppkeys/localhost.priv'
+[ YES ] Check public key: OK at '/var/cfengine/ppkeys/localhost.pub'
+...
+```
+
+```command
+docker exec -it cfengine-demo-cfengine-host-1 bash -c "/usr/local/sbin/cf-agent --self-diagnostics"
+```
+```output
+...
+[ YES ] Check that agent is bootstrapped: 192.168.16.2
+[ NO  ] Check if agent is acting as a policy server: Not acting as a policy server
+[ YES ] Check private key: OK at '/var/cfengine/ppkeys/localhost.priv'
+[ YES ] Check public key: OK at '/var/cfengine/ppkeys/localhost.pub'
```

Comment:
here I also get a lot of NOs

```
[ NO  ] Check persistent classes DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/cf_state.lmdb'
[ NO  ] Check checksums DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/checksum_digests.lmdb'
[ NO  ] Check observations DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/cf_observations.lmdb'
[ NO  ] Check file stats DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/stats.lmdb'
[ NO  ] Check locks DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/cf_lock.lmdb'
[ NO  ] Check performance DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/performance.lmdb'
[ NO  ] Check lastseen DB: Unable to diagnose LMDB file (not implemented) for '/var/cfengine/state/cf_lastseen.lmdb'
```

Reply:
@olehermanse is this expected? I haven't ever used this feature.

Reply:
@craigcomstock no, that's not expected, it is implemented. Here's from my hub:

```
$ cf-check diagnose
    info: No filenames specified, defaulting to .lmdb files in /var/cfengine/state
    info: Status of '/var/cfengine/state/cf_lock.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/packages_installed_apt_get.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/nova_track.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/cf_state.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/nova_measures.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/performance.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/cf_lastseen.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/cf_observations.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/packages_updates_apt_get.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/nova_agent_execution.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/cf_changes.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/history.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/nova_cookies.lmdb': OK [0% usage]
    info: Status of '/var/cfengine/state/nova_static.lmdb': OK [0% usage]
    info: All 14 databases healthy
```

Reply:
@craigcomstock can you make a reproducer and put it in a ticket?

Reply:
@craigcomstock ah, I see now. This is cf-agent self diagnostic. Yep, we don't often use that.

https://northerntech.atlassian.net/browse/CFE-4472